File: frontend/user-web/src/dtos/community-dto.ts
// --- Request DTOs ---

export interface CommunitySetting {
    is_private: boolean;              // Private community (only approved members can view)
    post_require_approval: boolean;   // Posts need moderator approval
    join_require_approval: boolean;   // Join requests need moderator approval
    max_post_length: number;          // Maximum post length
}

export interface Moderator {
    user_id: string;
    username: string;
    avatar?: {
        url: string;
        public_id?: string;
    };
    is_active?: boolean;
}

export interface CommunityRule {
    title: string;
    description: string;
}

export interface CreateCommunityRequest {
    name: string;
    description?: string;
    avatar?: string;
    banner?: string;
    setting?: CommunitySetting;
    rules?: CommunityRule[];
    moderators?: Moderator[];
    creator_name?: string;
    creator_avatar?: string;
    is_18_plus?: boolean;
}

export interface UpdateCommunityRequest {
    id: string;
    description?: string;
    avatar?: string;
    banner?: string;
    setting?: CommunitySetting;
    rules?: CommunityRule[];
}

export interface ModeratorDTO {
    id: string;
    username: string;
}

export interface AddModeratorRequest {
    id: string;
    added_moderator: string[]; // Array of user IDs
}

export interface RemoveModeratorRequest {
    id: string;
    removed_moderator: string[]; // Array of user IDs
}

export interface BanUserRequest {
    community_id: string;
    user_id: string;
    type: string; // "ban" or "mute"
    reason: string;
    length_days: number;
}

export interface UnbanUserRequest {
    community_id: string;
    user_id: string;
}

export interface UserResponse {
    id: string;
    username: string;
    email?: string;
    avatar?: string;
    banner?: string;
    bio?: string;
    created_at?: string;
    profile?: {
        avatar?: {
            url: string;
            public_id?: string;
        };
        cover?: {
            url: string;
            public_id?: string;
        };
        bio?: string;
    };
}

// --- Response DTOs ---

export interface PaginationInfo {
    current_page: number;
    total_pages: number;
    total_items: number;
    page_size: number;
}

export interface PaginatedCommunitiesResponse {
    communities: CommunityResponse[];
    pagination: PaginationInfo;
}

export interface CommunityResponse {
    id: string;
    name: string;
    description: string;
    avatar: string;
    banner: string;
    setting: CommunitySetting;
    rules?: CommunityRule[];
    moderators: Moderator[];
    post_count: number;
    member_count: number;
    create_by_id?: string;
    create_by_name?: string;
    create_by_avatar?: string;
    is_18_plus?: boolean;
    created_at?: string;
}

File: backend/internal/dto/community_dto.go
package dto

import (
	"time"

	"github.com/giakiet05/lkforum/internal/model"
)

type CreateCommunityRequest struct {
	Name          string                 `json:"name" binding:"required,min=3,max=50"`
	Description   *string                `json:"description,omitempty" binding:"max=500"`
	Avatar        *string                `json:"avatar,omitempty"`
	Banner        *string                `json:"banner,omitempty"`
	Setting       model.CommunitySetting `json:"setting,omitempty"`
	Rules         []model.CommunityRule  `json:"rules,omitempty"`
	Moderators    []model.Moderator      `json:"moderators,omitempty"`
	CreatorName   string                 `json:"creator_name,omitempty"`
	CreatorAvatar string                 `json:"creator_avatar,omitempty"`
	Is18Plus      bool                   `json:"is_18_plus"`
}

type UpdateCommunityRequest struct {
	CommunityID string                  `json:"id" binding:"required"`
	Description *string                 `json:"description,omitempty"`
	Avatar      *string                 `json:"avatar,omitempty"`
	Banner      *string                 `json:"banner,omitempty"`
	Setting     *model.CommunitySetting `json:"setting,omitempty"`
	Rules       *[]model.CommunityRule  `json:"rules,omitempty"`
}

type AddModeratorRequest struct {
	CommunityID    string   `json:"id" binding:"required"`
	AddedModerator []string `json:"added_moderator" binding:"required"`
}

type RemoveModeratorRequest struct {
	CommunityID      string   `json:"id" binding:"required"`
	RemovedModerator []string `json:"removed_moderator" binding:"required"`
}

type CommunityBanPostRequest struct {
	CommunityID string  `json:"community_id" binding:"required"`
	PostID      string  `json:"post_id" binding:"required"`
	Reason      *string `json:"reason,omitempty" binding:"max=500"`
}

type CommunityUnbanPostRequest struct {
	CommunityID string `json:"community_id" binding:"required"`
	PostID      string `json:"post_id" binding:"required"`
}

type CommunityBanUserRequest struct {
	CommunityID string `json:"community_id" binding:"required"`
	UserID      string `json:"user_id" binding:"required"`
	Type        string `json:"type" binding:"required"`
	Reason      string `json:"reason" binding:"max=500"`
	LengthDays  int    `json:"length_days" binding:"required"`
}

type UnbanUserRequest struct {
	CommunityID string `json:"community_id" binding:"required"`
	UserID      string `json:"user_id" binding:"required"`
}

type ModeratePostRequest struct {
	Approve bool    `json:"approve"`
	Reason  *string `json:"reason,omitempty"`
}

type CommunityResponse struct {
	ID             string                 `json:"id"`
	Name           string                 `json:"name"`
	Description    *string                `json:"description"`
	Avatar         *string                `json:"avatar"`
	Banner         *string                `json:"banner"`
	Setting        model.CommunitySetting `json:"setting"`
	Rules          []model.CommunityRule  `json:"rules"`
	Moderators     []model.Moderator      `json:"moderators"`
	PostCount      int64                  `json:"post_count"`
	MemberCount    int64                  `json:"member_count"`
	Is18Plus       bool                   `json:"is_18_plus"`
	CreateByID     string                 `json:"create_by_id,omitempty"`
	CreateByName   string                 `json:"create_by_name,omitempty"`
	CreateByAvatar string                 `json:"create_by_avatar,omitempty"`
	CreateAt       time.Time              `json:"create_at,omitempty"`
}

func FromCommunities(communities []*model.Community) []*CommunityResponse {
	var communityResponses []*CommunityResponse
	for _, community := range communities {
		communityResponses = append(communityResponses, FromCommunity(community))
	}
	return communityResponses
}

func FromCommunity(community *model.Community) *CommunityResponse {
	return &CommunityResponse{
		ID:             community.ID.Hex(),
		Name:           community.Name,
		Description:    community.Description,
		Avatar:         community.Avatar,
		Banner:         community.Banner,
		Setting:        community.Setting,
		Rules:          community.Rules,
		Moderators:     community.Moderators,
		PostCount:      community.PostCount,
		MemberCount:    community.MemberCount,
		Is18Plus:       community.Is18Plus,
		CreateByID:     community.CreateByID.Hex(),
		CreateByName:   community.CreateByName,
		CreateByAvatar: community.CreateByAvatar,
		CreateAt:       community.CreateAt,
	}
}
